docs: Split cartesian and next ADRs

romanc opened the pull request:
## Description

As discussed in the last GT4Py community meeting, we'd like to split the ADRs for cartesian and next in two subfolder for a clear separation.

## Requirements

- [ ] All fixes and/or new features come with corresponding tests.
  N/A
- [x] Important design decisions have been documented in the appropriate ADR inside the [docs/development/ADRs/](docs/development/ADRs/Index.md) folder.

romanc commented:
cscs-ci run

FlorianDeconinck reviewed (approved):
LGTM

Cartesian will definitely be an "after-the-fact" write up rather than a careful design like next was. I think that the ADRs will be more straightforward and simpler in that way

romanc commented:
I'm going to merge this in 24h unless someone comments otherwise in that window...

havogt reviewed (changes requested):
I favor a single template, but didn't have time to discuss it on our side.

romanc commented:
Thanks for letting us know. We keep this open as long as it needs.

egparedes commented:
I like the idea of using ADRs in `cartesian` but keeping them separated from `next`, but I have a couple of comments about the current PR.

First, I would also prefer to have a single template for both but I would like keep the current structure for compatibility with the existing ADRs. Regarding your comments about the current template:

> What I didn't like (and thus changed) are the following points
>   -  I didn't see any usage of tags (or at least it wasn't explained which tags to use), so I removed them

I agree. We can remove the tags.

>   -  I removed the metadata information authors, created, updated because these are available from git

I disagree here. Of course it's easy to get the commit information from git but keeping the main points in the document itself directly communicates which people was involved in the discussion, when it started and if/when was corrected for the last time, which I think it's important.  The git history may have smaller commits fixing typos or other minor changes from people which were not really involved in the discussion.

>   -  I removed the metadata status because I want the default status to be valid. Anything superseded/unused/deprecated is moved to the archived/ folder (as defined in the README) such that it is still easily available, but people clearly see that it is not valid anymore.

I agree about moving the not-longer active ADRs to an `archived/` folder, but I would still like to keep the status field in the document for clarity. It shouldn't be a big deal since this field would only change once if at all during the lifetime of a document.

>    - And since I had a separate one anyway, some minor wording changes

I'm fine with any rewording of the template which makes it easier to understand.

Another change I would propose in this PR is to move the GDPs to the archived folder inside the cartesian ADRs, explaining that they a older and overblown version of and ADR, and they are kept there for historical purposes.

romanc commented:
I'd like to keep two separate templates because I think writing tech docs (in the form of ADRs or whatever) should be in the competence of the team working on the code. And the teams working on gt4py-cartsian and gt4py-next are de-facto not overlapping (anymore). I think there needs to be room for difference between cartesian and next. And it's not that this is a new concept. For example, cartesian and next both have an optional dependency on DaCe and the version in next differs from the version in cartesian. And that's fine.

In general, I think the argument of "consistency with older ADRs" isn't valid because there are no older ADRs for gt4py-cartesian. If I understand correctly, the (outdated) tech docs for gt4py-cartesian are the GDPs, which have been actively stopped, leaving gt4py-cartesian currently without any tech docs. So I'd argue that whatever we start now doesn't need to be consistent with a previous format because there is no previous format. For me, writing tech docs in the form of ADRs is already an attempt to not deviate too far between gt4py-carteisan and -next.

> > * I removed the metadata information authors, created, updated because these are available from git
> 
> I disagree here. Of course it's easy to get the commit information from git but keeping the main points in the document itself directly communicates which people was involved in the discussion, when it started and if/when was corrected for the last time, which I think it's important. The git history may have smaller commits fixing typos or other minor changes from people which were not really involved in the discussion.

Then we have different opinions here and I think that's fine. For example, I think these tech docs should be about the trade-offs and ideas behind a feature and not about who was involved in the discussions. As long as we can have two templates, we don't have to argue and/or find a compromise. 
 
> > * I removed the metadata status because I want the default status to be valid. Anything superseded/unused/deprecated is moved to the archived/ folder (as defined in the README) such that it is still easily available, but people clearly see that it is not valid anymore.
> 
> I agree about moving the not-longer active ADRs to an `archived/` folder, but I would still like to keep the status field in the document for clarity. It shouldn't be a big deal since this field would only change once if at all during the lifetime of a document.

What I would like is to start reading the ADR document and immediately get to the content without having to skip the block of meta information. So for me, having the field in the first place is not useful information, especially if it only changes once. `archived/README.md` has clear instructions on how to link superseded ADRs to new ones.

As mentioned above, I think we should allow a bit of flexibility here between next and cartesian.

> Another change I would propose in this PR is to move the GDPs to the archived folder inside the cartesian ADRs, explaining that they a older and overblown version of and ADR, and they are kept there for historical purposes.

Happy to do this in a follow-up PR.

romanc commented:
Alright, thanks for the discussion in this month's community meeting. Let's have a shared ADR template for cartesian and next ADRs. The shared template currently

- has no tags anymore as agreed to by Enrique
- keeps the meta information block with with `status`, `authors`, `created` and `updated` fields
- has a "why statement" with two sentences because the sentence was getting really long
- rewords some of the text. Mostly I think we don't need [ square ] brackets around text to understand that it's a placeholder.

For the discussion about having an `archived/` folder: I kept the one for cartesian and didn't touch existing ADRs of next. I'm happy to do a follow-up PR for next if you find that useful (or you do that yourself).

For the discussion about moving the GDPs, I'd like to do that in a follow-up PR.

egparedes reviewed (approved):
Looks good to me, only have an optional suggestion to update the `next` README. Thanks for the effort to keep the documentation aligned between cartesian and next.

romanc commented on docs/development/ADRs/next/README.md:
```diff
@@ -0,0 +1,69 @@
+
+This document contains links to all _Architecture Decision Record_ (ADR) documents written in this project. Existing ADRs should convey the rationale for the design decisions in the current codebase and their associated trade-offs.
+
+## Writing ADRs
```
sure - happy to add this here :+1:

romanc commented:
Updated as proposed. The GPU tests are failing since CSCS-CI won't be able to allocate nodes this week. I'll re-run next week when CSCS-CI is scheduled to be operational again, unless someone with admin rights wants to push this side-stepping the CI.

egparedes commented:
It is safe to merge this PR since it doesn't contain code changes.
